packages/react-hooks/src/hooks/use-cart/use-cart.reducer.ts

```diff
@@ -29,7 +30,31 @@ const cartReducer = (
   switch (action.type) {
     case ADD_TO_CART: {
       const cart: CartItem[] = buildCart(state.cart, action.payload);
-
+      setCacheItem('cart', JSON.stringify(cart));
+      return {
+        ...state,
+        cart
+      };
+    }
+    case UPDATE_ITEM: {
+      const cart: CartItem[] = state.cart.map((item) => {
+        const payloadId =
+          'variant' in action.payload
+            ? action.payload.variant.id
+            : action.payload.id;
+        if (item.id === payloadId) {
+          Object.keys(item).forEach((key) => {
+            if (
+              action.payload[key] &&
+              key !== 'id' &&
+              item[key] !== action.payload[key]
+            ) {
+              item[key] = action.payload[key];
+            }
+          });
```

Comment:
This works, but it's probably good practice to take an immutable approach. This could be replaced by a single line:

```js
return { ...action.payload, id: item.id }
```

Reply:
Hey Nick, thanks for reviewing!

This is the approach I was going to use, but multiple problems presented themselves when I tried:

First off, typescript wouldn't let me use the payload as a cartItem

The bigger problem is that the cartItem is modified and doesn't have exactly the same fields as the payload, doing it as I did ensure that only an existing field can be modified, and it also ensures that some fields that shouldn't be modified aren't.

Reply:
> The bigger problem is that the cartItem is modified and doesn't have exactly the same fields as the payload, doing it as I did ensure that only an existing field can be modified, and it also ensures that some fields that shouldn't be modified aren't.

That makes sense! I'm good with that.

> typescript wouldn't let me use the payload as a cartItem

Ah, let me show you a trick for future reference. This gets at the concept of [Type Assertion](https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions). There are a couple of ways that you can assert what type something should be, when it could potentially be a lot of things. So in my example from before, here's how to let TypeScript know that we're enforcing that the payload is a `CartItem` in this context:

Example 1:

```typescript
return { ...(<CartItem>action.payload), id: item.id };
```

Example 2:

```typescript
return { ...(action.payload as CartItem), id: item.id };
```

Both are valid.

https://user-images.githubusercontent.com/5732000/110534345-f5d5d180-80ec-11eb-990f-522fcc6fd3d1.mov